api/src/main/java/org/pmiops/workbench/cdr/CdrDbConfig.java

```diff
@@ -0,0 +1,124 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.boot.autoconfigure.jdbc.DataSourceBuilder;
+import org.springframework.boot.orm.jpa.EntityManagerFactoryBuilder;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
+import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource;
+import org.springframework.orm.jpa.JpaTransactionManager;
+import org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean;
+import org.springframework.transaction.PlatformTransactionManager;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+
+@Configuration
+@EnableTransactionManagement
+@EnableJpaRepositories(
+    entityManagerFactoryRef = "cdrEntityManagerFactory",
+    transactionManagerRef = "cdrTransactionManager",
+    basePackages = { "org.pmiops.workbench.cdr" }
+)
+/**
+ * Spring configuration for connecting to a database with private or public CDR metadata based
+ * on the context of the current request. Applies to the model and DAO objects within this package.
+ */
+public class CdrDbConfig {
+
+  private static final String DB_DRIVER_CLASS_NAME_KEY = "spring.datasource.driver-class-name";
+  private static final String DB_URL_KEY = "spring.datasource.url";
+  private static final String DB_USER_KEY = "spring.datasource.username";
+  private static final String DB_PASSWORD_KEY = "spring.datasource.password";
+  private static final String WORKBENCH_DB_USER = "workbench";
+
+  public static class CdrDataSource extends AbstractRoutingDataSource {
+
+    @Autowired
+    public void CdrDataSource(CdrVersionDao cdrVersionDao) {
+      Map<String, String> envVariables = System.getenv();
+      String dbDriverClassName = envVariables.get(DB_DRIVER_CLASS_NAME_KEY);
+      String dbUser = envVariables.get(DB_USER_KEY);
+      String dbPassword = envVariables.get(DB_PASSWORD_KEY);
+      String originalDbUrl = envVariables.get(DB_URL_KEY);
+      boolean isWorkbenchDbUser = isWorkbenchDbUser(dbUser);
+
+      // Build a map of CDR version ID -> DataSource for use later, based on all the entries in the
+      // cdr_version table. Note that if new CDR versions are inserted, we need to restart the
+      // server in order for it to be used.
+      // TODO: find a way to make sure CDR versions aren't shown in the UI until they are in use by
+      // all servers.
+      Map<Object, Object> cdrVersionDataSourceMap = new HashMap<>();
+      for (CdrVersion cdrVersion : cdrVersionDao.findAll()) {
+        // The database name used for a given CDR version is depends on whether this is the
+        // workbench database user or data browser user.
+        String dbName = isWorkbenchDbUser ? cdrVersion.getCdrDbName() : cdrVersion.getPublicDbName();
```

Comment:
Since we are having a separate public api build would this just have a different config var with db name ? Or this line of code is changed in the public version?  I'm not sure how the public api is going to be implemented and what code files will be different and what files will be shared...

Reply:
One of my comments i don't see. But what i mean here is , in each api, it will either be public user or workbench user right ? So can cdrDBName  just be specified in a config for each api, so we don't need to do this check at all? or getCdrDbName() method can just be different in each api and return the right name? ... Know what i mean?

Reply:
I'm not sure if we will need config in public-api. At the moment, public-api doesn't need to use workbench DB for anything at all, and its DB user might not have permission to use it.

If we do need it to have config, we could talk about whether it's the same config or different, and what DB that lives in.

We could have a different environment variable set indicating the name of the DB to use for CDR metadata, if you prefer. I just figured the DB users were different so (perhaps hackily) we could just use that to determine what DB to access.

Reply:
Yeah, I'll add another setting for cdr DB name. That's fine.

Reply:
So I guess I should have said: we don't want to hardcode CDR db name in the properties for the server... the idea is there's a different CDR db name per CDR version, and you can switch between them.

We *could* have an explicit setting which is "use the public CDR db" or not. I'm not sure if that's better or worse than what I'm doing here, though -- we need to make sure we always use the public CDR db with the data browser user. WDYT?